### ISSUE TITLE
Can you support AOSONG ACD3100?

### ISSUE THREAD

**Issue by lorna1** [RobTillaart/ACD10#11]
Product url :http://www.aosong.com/products-220.html.
I tried to use the library code of ACD10 to connect ACD3100 (because it is the same protocol and the same interactive instruction from the official website document), but the 0x42 address isConnect() method has been output 0, and the example address from the official document is 0x54, and the connection cannot be adjusted. The chip I use is arduino 8266, and the pin uses VIN(5V)->pin4,d1->pin2,d2->pin1,GND->pin3.

**Comment by RobTillaart** [comment]
Thanks for your question,
Will look into it this week. If they are as similar as you describe it should be not too complex.

**Comment by RobTillaart** [comment]
Had a quick look and the datasheets look indeed pretty similar.
Will create a repo as placeholder.

**Comment by RobTillaart** [comment]
Created https://github.com/RobTillaart/ACD3100
The initial version is a renamed version of ACD10 repo and functions as a start point for development.

Please open issues there if things not work as expected.

**Comment by RobTillaart** [comment]
> but the 0x42 address isConnect() method has been output 0

@lorna1 
Can you run an I2C scanner to see which address is actually used?

e.g. one from https://github.com/RobTillaart/I2C_SCANNER

**Comment by RobTillaart** [comment]
@lorna1 

> The chip I use is arduino 8266, and the pin uses VIN(5V)->pin4,d1->pin2,d2->pin1,GND->pin3.

Do you have pull up resistors (2.2 KOhm) on the SDA and SCL lines ?

**Comment by lorna1** [comment]
> > but the 0x42 address isConnect() method has been output 0
> 
> @lorna1 Can you run an I2C scanner to see which address is actually used?
> 
> e.g. one from https://github.com/RobTillaart/I2C_SCANNER

I used the code above to scan all the addresses, with no results.
However, I found a phenomenon that if the GND pin is separated briefly, the isConnected() method will output 1 for a short time. And then on the next loop it goes back to 0.


```cpp
#include <Wire.h>;
 
void setup(){
  Wire.begin();
  Serial.begin(9600);
  Serial.println("\nI2C Scanner");
}

void loop(){
  byte error, address;
  int nDevices;
  Serial.println("Scanning...");
  nDevices = 0;
  for (address = 1; address < 127; address++ ){
    // The i2c_scanner uses the return value of
    // the Write.endTransmisstion to see if
    // a device did acknowledge to the address.
    Wire.beginTransmission(address);
    error = Wire.endTransmission();
    if (error == 0){
      Serial.print("I2C device found at address 0x");
      if (address < 16)
        Serial.print("0");
      Serial.print(address, HEX);
      Serial.println(" !");
      nDevices++;
    }else if (error == 4){
     
    }
  }
  if (nDevices == 0)
    Serial.println("No I2C devices found\n");
  else
    Serial.println("done\n");
  delay(5000); // wait 5 seconds for next scan
}
```

(updated for syntax highlighting)

**Comment by lorna1** [comment]
> > but the 0x42 address isConnect() method has been output 0
> 
> @lorna1 Can you run an I2C scanner to see which address is actually used?
> 
> e.g. one from https://github.com/RobTillaart/I2C_SCANNER

I tried the address 0X54 on the document and there was no normal connection.

**Comment by RobTillaart** [comment]
> However, I found a phenomenon that if the GND pin is separated briefly, the isConnected() method will output 1 for a short time. And then on the next loop it goes back to 0.

Cannot explain right now, have other tasks to do so to be continued later.

> I tried the address 0X54 on the document and there was no normal connection.

I2C addresses shift 1 bit and depending on read or write the add a 0 or 1.
If you look at the datasheet you see under address sometimes 0x54 and sometimes 0x55. 
The difference is the RW bit.

Please test with 0x2A too (which is shifted version of 0x54)

**Comment by lorna1** [comment]
> https://github.com/RobTillaart/ACD3100

Thank you very much for your reply

**Comment by RobTillaart** [comment]
@lorna1 

FYI, just released the 0.1.0 version, so we have a base to work from.

**Comment by lorna1** [comment]
I used the development board of ESP8266 CH340, and tried ESP8266 CP2102,github address:
https://github.com/nodemcu/nodemcu-devkit?tab=readme-ov-file

**Comment by lorna1** [comment]
The following code (ACD10.c ACD10.h iic.c iic.h)st development board official website demo.

```
#include "ACD10.h"
#include "iic.h"

extern	u32 CO2Data;


u32 ACD10_Read_Data(ComTypeDef IIC_COM)
{
	u8 Byte[6],CRC_Byte[3];
	
	
	Init_IIC_Sensor_Port(IIC_COM);
	
	
	IIC_Start(IIC_COM);
	IIC_WR_Byte(0x54,IIC_COM);
	if(Receive_ACK(IIC_COM)==1)return 1;
	IIC_WR_Byte(0x03,IIC_COM);
	if(Receive_ACK(IIC_COM)==1)return 1;
	IIC_WR_Byte(0x00,IIC_COM);
	if(Receive_ACK(IIC_COM)==1)return 1;
	IIC_Stop(IIC_COM);
	
	HAL_Delay(1);
	
	IIC_Start(IIC_COM);
	IIC_WR_Byte(0x55,IIC_COM);
	if(Receive_ACK(IIC_COM)==1)return 1;
	Byte[0]=IIC_RD_Byte(IIC_COM);
	Send_ACK(IIC_COM);
	Byte[1]=IIC_RD_Byte(IIC_COM);
	Send_ACK(IIC_COM);
	CRC_Byte[0]=IIC_RD_Byte(IIC_COM);
	Send_ACK(IIC_COM);
	Byte[2]=IIC_RD_Byte(IIC_COM);
	Send_ACK(IIC_COM);
	Byte[3]=IIC_RD_Byte(IIC_COM);
	Send_ACK(IIC_COM);
	CRC_Byte[1]=IIC_RD_Byte(IIC_COM);
	Send_ACK(IIC_COM);
	Byte[4]=IIC_RD_Byte(IIC_COM);
	Send_ACK(IIC_COM);
	Byte[5]=IIC_RD_Byte(IIC_COM);
	Send_ACK(IIC_COM);
	CRC_Byte[2]=IIC_RD_Byte(IIC_COM);
	Send_NOT_ACK(IIC_COM);
	IIC_Stop(IIC_COM);
	
	if((Calc_CRC8(Byte,2)!=CRC_Byte[0])||(Calc_CRC8(Byte+2,2)!=CRC_Byte[1])||(Calc_CRC8(Byte+4,2)!=CRC_Byte[2]))
	return 2;//CRCУ�����
	
	CO2Data=(CO2Data|Byte[0])<<8;
	CO2Data=(CO2Data|Byte[1])<<8;
	CO2Data=(CO2Data|Byte[2])<<8;
	CO2Data=(CO2Data|Byte[3]);
	
	return 0;
	
	
	
}
```

```
#ifndef __ACD10_H_
#define __ACD10_H_

#include "iic.h"


u32 ACD10_Read_Data(ComTypeDef IIC_COM);






















#endif
```


```
#include "iic.h" 
#include "main.h"
#include "delay.h"




ComTypeDef IIC_COM[5];

void COM_Config() //���ӿ�IIC���ţ��˿�����
{
	
IIC_COM[0].SDA_PIN=IIC_SDA_0_Pin;
IIC_COM[0].SCL_PIN=IIC_SCL_0_Pin;
IIC_COM[0].SDA_PORT=IIC_SDA_0_GPIO_Port;
IIC_COM[0].SCL_PORT=IIC_SCL_0_GPIO_Port;
	
IIC_COM[1].SDA_PIN=IIC_SDA_1_Pin;
IIC_COM[1].SCL_PIN=IIC_SCL_12_Pin;
IIC_COM[1].SDA_PORT=IIC_SDA_1_GPIO_Port;
IIC_COM[1].SCL_PORT=IIC_SCL_12_GPIO_Port;
	
IIC_COM[2].SDA_PIN=IIC_SDA_2_Pin;
IIC_COM[2].SCL_PIN=IIC_SCL_12_Pin;
IIC_COM[2].SDA_PORT=IIC_SDA_2_GPIO_Port;
IIC_COM[2].SCL_PORT=IIC_SCL_12_GPIO_Port;
	
IIC_COM[3].SDA_PIN=IIC_SDA_3_Pin;
IIC_COM[3].SCL_PIN=IIC_SCL_34_Pin;
IIC_COM[3].SDA_PORT=IIC_SDA_3_GPIO_Port;
IIC_COM[3].SCL_PORT=IIC_SCL_34_GPIO_Port;
	
IIC_COM[4].SDA_PIN=IIC_SDA_4_Pin;
IIC_COM[4].SCL_PIN=IIC_SCL_34_Pin;
IIC_COM[4].SDA_PORT=IIC_SDA_4_GPIO_Port;
IIC_COM[4].SCL_PORT=IIC_SCL_34_GPIO_Port;
		

}







void SDA_Pin_Output_High(ComTypeDef IIC_COM)   //��SDA����Ϊ��©��� ��������Ϊ�ߵ�ƽ
{
	GPIO_InitTypeDef GPIO_InitStruct;
  GPIO_InitStruct.Pin = IIC_COM.SDA_PIN;
  GPIO_InitStruct.Mode = GPIO_MODE_OUTPUT_OD;
  GPIO_InitStruct.Pull = GPIO_PULLUP;
  GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_HIGH;
  HAL_GPIO_Init(IIC_COM.SDA_PORT, &GPIO_InitStruct);
	HAL_GPIO_WritePin(IIC_COM.SDA_PORT, IIC_COM.SDA_PIN, GPIO_PIN_SET);
}

void SDA_Pin_Output_Low(ComTypeDef IIC_COM)  //��SDA����Ϊ��©���  ������Ϊ�͵�ƽ
{

	GPIO_InitTypeDef GPIO_InitStruct;
  GPIO_InitStruct.Pin = IIC_COM.SDA_PIN;
  GPIO_InitStruct.Mode = GPIO_MODE_OUTPUT_OD;
  GPIO_InitStruct.Pull = GPIO_PULLUP;
  GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_HIGH;
  HAL_GPIO_Init(IIC_COM.SDA_PORT, &GPIO_InitStruct);
	HAL_GPIO_WritePin(IIC_COM.SDA_PORT, IIC_COM.SDA_PIN, GPIO_PIN_RESET);
}

void SDA_Pin_INPUT(ComTypeDef IIC_COM)  //SDA����Ϊ��������
{

	GPIO_InitTypeDef GPIO_InitStruct;
  GPIO_InitStruct.Pin = IIC_COM.SDA_PIN;
  GPIO_InitStruct.Mode = GPIO_MODE_INPUT;
  GPIO_InitStruct.Pull = GPIO_PULLUP;
  GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_HIGH;
  HAL_GPIO_Init(IIC_COM.SDA_PORT, &GPIO_InitStruct);

}

void SCL_Pin_Output_High(ComTypeDef IIC_COM) //SCL����Ϊ��©�����������Ϊ�ߵ�ƽ
{
	GPIO_InitTypeDef GPIO_InitStruct;
  GPIO_InitStruct.Pin = IIC_COM.SCL_PIN;
  GPIO_InitStruct.Mode = GPIO_MODE_OUTPUT_OD;
  GPIO_InitStruct.Pull = GPIO_PULLUP;
  GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_HIGH;
  HAL_GPIO_Init(IIC_COM.SCL_PORT, &GPIO_InitStruct);
	HAL_GPIO_WritePin(IIC_COM.SCL_PORT, IIC_COM.SCL_PIN, GPIO_PIN_SET);
}

void SCL_Pin_Output_Low(ComTypeDef IIC_COM) //SCL����Ϊ��©�����������͵�ƽ
{
	GPIO_InitTypeDef GPIO_InitStruct;
  GPIO_InitStruct.Pin = IIC_COM.SCL_PIN;
  GPIO_InitStruct.Mode = GPIO_MODE_OUTPUT_OD;
  GPIO_InitStruct.Pull = GPIO_PULLUP;
  GPIO_InitStruct.Speed = GPIO_SPEED_FREQ_HIGH;
  HAL_GPIO_Init(IIC_COM.SCL_PORT, &GPIO_InitStruct);
	HAL_GPIO_WritePin(IIC_COM.SCL_PORT, IIC_COM.SCL_PIN, GPIO_PIN_RESET);
}

void Init_IIC_Sensor_Port(ComTypeDef IIC_COM) //��ʼ��IIC�ӿ�,���Ϊ�ߵ�ƽ
{	
SDA_Pin_Output_High(IIC_COM);
SCL_Pin_Output_High(IIC_COM);
}
void IIC_Start(ComTypeDef IIC_COM)		 //IIC��������START�ź�
{
	SDA_Pin_Output_High(IIC_COM);
	IIC_delay;
	SCL_Pin_Output_High(IIC_COM);
	IIC_delay;
	SDA_Pin_Output_Low(IIC_COM);
	IIC_delay;
	SCL_Pin_Output_Low(IIC_COM);
	IIC_delay;   
}


void IIC_WR_Byte(uint8_t Byte,ComTypeDef IIC_COM) //IICдһ���ֽ�
{
	uint8_t Data,N,i;	
	Data=Byte;
	i = 0x80;
	for(N=0;N<8;N++)
	{
		SCL_Pin_Output_Low(IIC_COM); 
		IIC_delay;	
		if(i&Data)
		{
			SDA_Pin_Output_High(IIC_COM);
		}
		else
		{
			SDA_Pin_Output_Low(IIC_COM);
		}	
			
    SCL_Pin_Output_High(IIC_COM);
		IIC_delay;
		Data <<= 1;
		 
	}
	SCL_Pin_Output_Low(IIC_COM);
	IIC_delay;   
	SDA_Pin_INPUT(IIC_COM);
	IIC_delay;
	
}	


uint8_t IIC_RD_Byte(ComTypeDef IIC_COM)//IIC��ȡһ���ֽ�
{
	uint8_t Byte,i,a;
	Byte = 0;
	SCL_Pin_Output_Low(IIC_COM);
	SDA_Pin_INPUT(IIC_COM);
	IIC_delay;	
	for(i=0;i<8;i++)
	{
    SCL_Pin_Output_High(IIC_COM);		
		IIC_delay;
		a=0;
		if(HAL_GPIO_ReadPin(IIC_COM.SDA_PORT, IIC_COM.SDA_PIN)) a=1;
		Byte = (Byte<<1)|a;
		SCL_Pin_Output_Low(IIC_COM);
		IIC_delay;
	}
  SDA_Pin_INPUT(IIC_COM);
	IIC_delay;	
	return Byte;
}


uint8_t Receive_ACK(ComTypeDef IIC_COM)   //����ACK,����Ӧ�𣬷���1
{
	uint16_t CNT;
	CNT = 0;
	SCL_Pin_Output_Low(IIC_COM);	
	SDA_Pin_INPUT(IIC_COM);
	IIC_delay;	
	SCL_Pin_Output_High(IIC_COM);	
	IIC_delay;	
	while((HAL_GPIO_ReadPin(IIC_COM.SDA_PORT, IIC_COM.SDA_PIN))  && CNT < 100) 
	CNT++;
	if(CNT == 100)
	{
		return 1;//��Ӧ��
	}
 	SCL_Pin_Output_Low(IIC_COM);	
	IIC_delay;	
	return 0;
}

void Send_ACK(ComTypeDef IIC_COM)		  //�����ظ�ACK�ź�
{
	SCL_Pin_Output_Low(IIC_COM);	
	IIC_delay;	
	SDA_Pin_Output_Low(IIC_COM);
	IIC_delay;	
	SCL_Pin_Output_High(IIC_COM);	
	IIC_delay;
	SCL_Pin_Output_Low(IIC_COM);	
	IIC_delay;
	SDA_Pin_INPUT(IIC_COM);
	IIC_delay;
}

void Send_NOT_ACK(ComTypeDef IIC_COM)	//�������ظ�ACK
{
	SCL_Pin_Output_Low(IIC_COM);	
	IIC_delay;
	SDA_Pin_Output_High(IIC_COM);
	IIC_delay;
	SCL_Pin_Output_High(IIC_COM);	
	IIC_delay;		
	SCL_Pin_Output_Low(IIC_COM);	
	IIC_delay;
    SDA_Pin_Output_Low(IIC_COM);
	IIC_delay;
}

void IIC_Stop(ComTypeDef IIC_COM)	  //һ��Э�����
{
	SDA_Pin_Output_Low(IIC_COM);
	IIC_delay;
	SCL_Pin_Output_High(IIC_COM);	
	IIC_delay;
	SDA_Pin_Output_High(IIC_COM);
	IIC_delay;
}






u8 Calc_CRC8(u8 *dat, u8 Num)
{
	u8 i,byte,crc=0xFF;
	for(byte=0; byte<Num; byte++)
	{
		crc^=(dat[byte]);
		for(i=0;i<8;i++)
		{
			if(crc & 0x80) crc=(crc<<1)^0x31;
			else crc=(crc<<1);
		}
	}
	return crc;
}
```

```
#ifndef _AHT20_DEMO_
#define _AHT20_DEMO_

#include "stm32g0xx.h"  
#include "delay.h"


#define IIC_delay delay_us(20)




typedef struct
{
  uint16_t SDA_PIN;        
	uint16_t SCL_PIN;
	GPIO_TypeDef * SDA_PORT;
	GPIO_TypeDef * SCL_PORT;
} ComTypeDef;








void COM_Config(void);

void SDA_Pin_Output_High(ComTypeDef IIC_COM);   //��SDA����Ϊ��©��� ��������Ϊ�ߵ�ƽ

void SDA_Pin_Output_Low(ComTypeDef IIC_COM);  //��SDA����Ϊ��©���  ������Ϊ�͵�ƽ

void SDA_Pin_INPUT(ComTypeDef IIC_COM);  //SDA����Ϊ��������

void SCL_Pin_Output_High(ComTypeDef IIC_COM); //SCL����ߵ�ƽ

void SCL_Pin_Output_Low(ComTypeDef IIC_COM); //SCL����͵�ƽ

void Init_IIC_Sensor_Port(ComTypeDef IIC_COM); //��ʼ��IIC�ӿ�,���Ϊ�ߵ�ƽ

void IIC_Start(ComTypeDef IIC_COM);		 //IIC��������START�ź�

void IIC_WR_Byte(uint8_t Byte,ComTypeDef IIC_COM); //IICдһ���ֽ�

uint8_t IIC_RD_Byte(ComTypeDef IIC_COM);//IIC��ȡһ���ֽ�

uint8_t Receive_ACK(ComTypeDef IIC_COM);   //����ACK

void Send_ACK(ComTypeDef IIC_COM);		  //�����ظ�ACK�ź�

void Send_NOT_ACK(ComTypeDef IIC_COM);	//�������ظ�ACK

void IIC_Stop(ComTypeDef IIC_COM);	  //һ��Э�����

u8 Calc_CRC8(u8 *dat, u8 Num);





#endif
```

**Comment by RobTillaart** [comment]
@lorna1 
>  The following code (ACD10.c ACD10.h iic.c iic.h)st development board official website demo.

OK, and does it work with the ACD3100 ? 

At the moment the first problem to tackle is why does the ACD3100 not show up in an I2C scanner?
If the I2C address cannot be seen, no code or library will be able to connect and read the sensor.

So step 1 is to check the hardware level.

Do you have pull up resistors on the SDA / SCL line? 
If so which value?

**Comment by RobTillaart** [comment]
![image](https://github.com/user-attachments/assets/a48ae3a6-c01f-46f0-9f26-92e38f955ebd)


Section 2.2 datasheet

_The sensor provides I2C and UART communication interfaces, which users can select according to their needs. The interface input level is compatible with 3V to 5V. The output level is internally pulled up to 5V. The sensor defaults (Pin 5 is suspended, with a built-in 4.7Kÿ pull-up resistor) to I2C communication; Pin 5 is connected to a low level._

@lorna1 
Can you explicitly connect pin5 of the ACD3100 to +5V

**Comment by RobTillaart** [comment]
![image](https://github.com/user-attachments/assets/6802b9ab-6ee0-4176-ad53-a0cea5cc84f7)

(above image from RandomNerdTutorials)

> The chip I use is arduino 8266, and the pin uses 
- VIN(5V)->pin 4,
- d1->pin 2,
- d2->pin 1,
- GND->pin 3.

So the pins look OK, but please add 5V to pin 5 too.

**Comment by lorna1** [comment]
> ![image](https://private-user-images.githubusercontent.com/462844/381073531-a48ae3a6-c01f-46f0-9f26-92e38f955ebd.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzAxOTY2MTYsIm5iZiI6MTczMDE5NjMxNiwicGF0aCI6Ii80NjI4NDQvMzgxMDczNTMxLWE0OGFlM2E2LWMwMWYtNDZmMC05ZjI2LTkyZTM4Zjk1NWViZC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMDI5JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTAyOVQxMDA1MTZaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT04NzUwY2ZmYjE0YjhkNmU4ZTU0ODJiMzY1MmIzZjE5ZThhMTlkZGE2OThmMDk1MWU3ZDU3OGZjN2EzMGEzY2Y4JlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.dvmHMG_FIQjx9Q1Y9xAKYlY9fkhiOk1xmb5y7Y5HX3E)
> 
> Section 2.2 datasheet
> 
> _The sensor provides I2C and UART communication interfaces, which users can select according to their needs. The interface input level is compatible with 3V to 5V. The output level is internally pulled up to 5V. The sensor defaults (Pin 5 is suspended, with a built-in 4.7Kÿ pull-up resistor) to I2C communication; Pin 5 is connected to a low level._
> 
> @lorna1 Can you explicitly connect pin5 of the ACD3100 to +5V

pin5 defaults to I2C communication, so my pin5 pin is not connected. I'll try powering Pin5 +5v for a try.

**Comment by lorna1** [comment]
> > but the 0x42 address isConnect() method has been output 0
> 
> @lorna1 Can you run an I2C scanner to see which address is actually used?
> 
> e.g. one from https://github.com/RobTillaart/I2C_SCANNER
If I get pin5, output
23:03:37.784 -> --- Scan started ---
23:03:37.784 -> No I2C devices found
23:03:37.820 -> --- Scan finished ---

If I pull pin5 it will print
23:03:57.856 -> --- Scan started ---
23:03:57.856 -> Unknow error at address 0x01
23:03:57.898 -> Unknow error at address 0x02
23:03:57.932 -> Unknow error at address 0x03
23:03:57.965 -> Unknow error at address 0x04
23:03:57.999 -> Unknow error at address 0x05
23:03:58.031 -> Unknow error at address 0x06
23:03:58.065 -> Unknow error at address 0x07
23:03:58.098 -> Unknow error at address 0x08
23:03:58.131 -> Unknow error at address 0x09
23:03:58.165 -> Unknow error at address 0x0A
23:03:58.197 -> Unknow error at address 0x0B
23:03:58.231 -> Unknow error at address 0x0C
23:03:58.231 -> Unknow error at address 0x0D
23:03:58.265 -> Unknow error at address 0x0E
23:04:57.318 -> No I2C devices found
23:04:57.357 -> --- Scan finished ---
![Uploading _cgi-bin_mmwebwx-bin_webwxgetmsgimg__&MsgID=2178543726987676054&skey=@crypt_1cd41551_acd00456bb88e5d7e362eddb22e6e657&mmweb_appid=wx_webfilehelper.jpg…]()

**Comment by lorna1** [comment]
![_cgi-bin_mmwebwx-bin_webwxgetmsgimg__ MsgID=2178543726987676054 skey=@crypt_1cd41551_acd00456bb88e5d7e362eddb22e6e657 mmweb_appid=wx_webfilehelper](https://github.com/user-attachments/assets/affc45aa-2990-4f9d-bf84-059527da6543)
![_cgi-bin_mmwebwx-bin_webwxgetmsgimg__ MsgID=4295501808115930912 skey=@crypt_1cd41551_acd00456bb88e5d7e362eddb22e6e657 mmweb_appid=wx_webfilehelper](https://github.com/user-attachments/assets/d54bf0ec-9bdc-4245-9282-0814e160d7c0)
![_cgi-bin_mmwebwx-bin_webwxgetmsgimg__ MsgID=4972929837428850373 skey=@crypt_1cd41551_acd00456bb88e5d7e362eddb22e6e657 mmweb_appid=wx_webfilehelper](https://github.com/user-attachments/assets/f404b524-acf9-42ea-ad7f-b7f581103710)
![_cgi-bin_mmwebwx-bin_webwxgetmsgimg__ MsgID=6369196837811049067 skey=@crypt_1cd41551_acd00456bb88e5d7e362eddb22e6e657 mmweb_appid=wx_webfilehelper](https://github.com/user-attachments/assets/c1b90809-9d89-4b8d-8121-fea6ec59ec94)

**Comment by RobTillaart** [comment]
The blue wire is now connected to 3V3, should be 5V.

**Comment by RobTillaart** [comment]
The remaining wires look OK.

**Comment by RobTillaart** [comment]
Can you post a photo of the sensor from both sides (TOP / BOTTOM)

**Comment by RobTillaart** [comment]
TOP view

![image](https://github.com/user-attachments/assets/e4178ead-cf8a-447a-8b6f-42e4b9540a4d)


BOTTOM view

![image](https://github.com/user-attachments/assets/f1be6faa-3a05-4708-8eda-eec5e23678b9)

**Comment by RobTillaart** [comment]
@lorna1 

Just an (awful) insight ...

I think the datasheet gives the TOP view and your photos show the BOTTOM view, 
That means in practice the pins are reversed, so the connections should be

|  pinname  |  meaning  |   wire    |
|:---------:|:---------:|:---------:|
|     5     |   5 V     |  orange   |
|     G     |   GND     |  black    |
|     T     |   SCL/TX  |  white    |
|     R     |   SDA/RX  |  yellow   |

Does your sensor have the characters ?

**Comment by lorna1** [comment]
The pin-graph in the document is actually mirrored, and the data can already be obtained by following the correct connection.The chip's LED lights started flashing.

I tested the sample program of connection, it is the address of the 0x2A, through mySensor.getCO2Concentration() can be read by CO2 readings for 1190.
![WechatIMG17](https://github.com/user-attachments/assets/cfdd9eb9-cf66-4e43-9c5d-c23e0182522e)
![WechatIMG16](https://github.com/user-attachments/assets/ba455c61-fedc-4bfd-9b71-d9c26ea4c2ae)

**Comment by RobTillaart** [comment]
> I tested the sample program of connection, it is the address of the 0x2A, through mySensor.getCO2Concentration() can be read by CO2 readings for 1190.

That is good to hear the sensor is working now! 
1190 is a reasonable CO2 for indoors I think.

Can you run **ACD3100_readSensor.ino** and post the output?  (or ACD10_readSensor.ino)

**Comment by lorna1** [comment]
I tried ACD10_readSensor.ino and ACD3100_readSensor.ino, where mySensor.requestReady() always outputs 0 and the code keeps loping indefinitely.

Serial.print(mySensor.requestReady());
08:44:54.797 -> 0
08:44:55.825 -> 0
08:44:56.824 -> 0

```cpp
void loop()
{
  //  prepare a request every 5 seconds.
  if (millis() - mySensor.lastRead() > 5000)  //  millis
  {
    mySensor.requestSensor();
  }
  Serial.print(mySensor.requestReady());
  Serial.println();
  //  if request has had enough time read the sensor.
  if (mySensor.requestReady())
  {
    mySensor.readSensor();
    Serial.print(mySensor.getCO2Concentration());
    Serial.print("\t");
    Serial.print(mySensor.getTemperature());
    Serial.println();
  }

  //  do other things here.
  delay(1000);
}

```

**Comment by RobTillaart** [comment]
Mmm, unexpected

We have to go through all demo sketches to see what is working or not.

Can you run ACD10_readSensorCode.ino?

**Comment by lorna1** [comment]
ACD10_readSensorCode.ino output
```
void loop()
{
  mySensor.readFirmwareVersion(&FWversion[0]);
  Serial.print("FIRMWARE: \t");
  Serial.println(FWversion);
  
  mySensor.readSensorCode(&sensorCode[0]);
  Serial.print("SENSORCODE: \t");
  Serial.println(sensorCode);
  delay(5000);
}

21:17:17.549 -> SENSORCODE: 	0188152356
21:17:17.549 -> FIRMWARE: 	AS V1.0.00
```

ACD10_readSensor_performance.ino
Code blocking at mySensor.requestReady()

**Comment by RobTillaart** [comment]
> ACD10_readSensorCode.ino output ...

So those look OK.

> ACD10_readSensor_performance.ino
Code blocking at mySensor.requestReady()

mmm.  (thinking sounds)

How long has it been running? 
It needs a few minutes to warm up and it might report not ready until after warming up. (3 minutes?)

**Comment by RobTillaart** [comment]
> Code blocking at mySensor.requestReady()

Can only return false if 
1. _requestStart is not set
2. millis() does not work.

Can you run this code to test hypothesis 2?

```cpp
void setup()
{
  Serial.begin(115200);
}

void loop()
{
  Serial.println(millis());
}
````

**Comment by RobTillaart** [comment]
Can you make a change in ACD3100.cpp

```cpp

int ACD3100::requestSensor()
{
  uint8_t buf[2] = { 0x03, 0x00 };
  Serial.println(_requestStart);
  _requestStart = millis();
  Serial.println(_requestStart);
  return _command(buf, 2);
}
````

and then run ACD3100_readSensor_performance.ino

**Comment by lorna1** [comment]
> > ACD10_readSensorCode.ino output ...
> 
> So those look OK.
> 
> > ACD10_readSensor_performance.ino
> > Code blocking at mySensor.requestReady()
> 
> mmm. (thinking sounds)
> 
> How long has it been running? It needs a few minutes to warm up and it might report not ready until after warming up. (3 minutes?)

Maybe it was a warm-up problem, I didn't pay attention to him for a few minutes, and came back later to see that the serial port log was winning normally.

21:29:39.934 -> done...
21:39:40.151 -> |  100000  |  13  |  6  |
21:39:40.340 -> |  200000  |  9  |  7  |
21:39:40.525 -> |  300000  |  9  |  7  |
21:39:40.711 -> |  400000  |  9  |  7  |
21:39:40.899 -> |  500000  |  9  |  7  |
21:39:41.103 -> |  600000  |  9  |  7  |
21:39:41.289 -> |  700000  |  13  |  7  |
21:39:41.477 -> |  800000  |  9  |  7  |

**Comment by RobTillaart** [comment]
OK, no need to change the ACD3100.cpp file as requested. 

Can you redo the ACD3100_demo.ino  and watch it for e.g. 10 minutes to see if it gets "unblocked" (normal mode).

**Comment by RobTillaart** [comment]
> 21:17:17.549 -> SENSORCODE: 	0188152356
21:17:17.549 -> FIRMWARE: 	AS V1.0.00

Have you checked your sensor if it mentions this number somewhere on a sticker or so?

**Comment by RobTillaart** [comment]
Going through the datasheet, I noticed that the ACD3100 does not support **SetCalibrationMode()**.
(the ACD10 has this function)

Made an issue for this.

**Comment by RobTillaart** [comment]
Will release ACD3100 library version 0.1.1 asap.

**Comment by RobTillaart** [comment]
ACD3100 released version 0.1.1

@lorna1 
All seems to work now in the ACD3100 library or did I miss something?

**Comment by lorna1** [comment]
![_cgi-bin_mmwebwx-bin_webwxgetmsgimg__ MsgID=2697560300853228719 skey=@crypt_1cd41551_66a49c0f6d2a605f99d21e5cb6c6bb9b mmweb_appid=wx_webfilehelper](https://github.com/user-attachments/assets/296d71a7-c0f8-44ab-b552-f8f385e6e6ca)

**Comment by lorna1** [comment]
```cpp
//
//    FILE: ACD3100_readSensor_performance.ino
//  AUTHOR: Rob Tillaart
// PURPOSE: test basic behaviour and performance
//     URL: https://github.com/RobTillaart/ACD3100


#include "ACD3100.h"


ACD3100 mySensor;


void test(uint32_t speed)
{
  Wire.setClock(speed);
  delay(100);

  uint32_t start = micros();
  mySensor.requestSensor();
  uint32_t duration1 = micros() - start;

  while(mySensor.requestReady() == false) delay(10);
  
  start = micros();
  mySensor.readSensor();
  uint32_t duration2 = micros() - start;

  Serial.print("|  ");
  Serial.print(speed);
  Serial.print("  |  ");
  Serial.print(duration1);
  Serial.print("  |  ");
  Serial.print(duration2);
  Serial.print("  |");
  Serial.println();
  //  reset I2C bus
  Wire.setClock(100000);
}


void setup()
{
  Serial.begin(115200);
  Serial.println();
  Serial.println(__FILE__);
  Serial.print("ACD3100_LIB_VERSION: ");
  Serial.println(ACD3100_LIB_VERSION);

  Wire.begin();
  mySensor.begin();

  Serial.println("\ndone...");
}


void loop()
{
  
  test(100000);
  test(200000);
  test(300000);
  test(400000);
  test(500000);
  test(600000);
  test(700000);
  test(800000);

  Serial.println("\ndone...");
  delay(600000);
}


//  -- END OF FILE --

```

Wait a long time for loop's test method to execute. I don't know if it's normal.

**Comment by RobTillaart** [comment]
The library has functions to test the warming up has been done. Have to check if the 120 seconds is in line with data sheet.

From readme.md

### PreHeat

PreHeat functions assume the sensor is (and stays) connected to power.

bool preHeatDone() returns true 120 seconds after constructor is called.

uint32_t preHeatMillisLeft() returns the time in milliseconds left before preHeat is complete.

**Comment by RobTillaart** [comment]
Final question: Can you post the output of this sketch?  (with the first five CO2 + temp readings or so)

it tests the preHeat functions.

```cpp
//
//    FILE: ACD3100_demo.ino
//  AUTHOR: Rob Tillaart
// PURPOSE: test basic behaviour and performance
//     URL: https://github.com/RobTillaart/ACD3100


#include "ACD3100.h"


ACD3100 mySensor;


void setup()
{
  Serial.begin(115200);
  Serial.println();
  Serial.println(__FILE__);
  Serial.print("ACD10_LIB_VERSION: ");
  Serial.println(ACD10_LIB_VERSION);

  Wire.begin();
  mySensor.begin();
  
  while(mySensor.preHeatDone() == false)
  {
    delay(10000);
    Serial.print("..");
    Serial.println(mySensor.preHeatMillisLeft()/1000);
  }
  Serial.println();
}


void loop()
{
  mySensor.requestSensor();
  while (mySensor.requestReady() == false) delay(10);
  mySensor.readSensor();
  Serial.print(mySensor.getCO2Concentration());
  Serial.print("\t");
  Serial.print(mySensor.getTemperature());
  Serial.println();

  delay(2000);
}


//  -- END OF FILE --
```

**Comment by RobTillaart** [comment]
@lorna1 

Also I want to understand the layout of the temperature data as the datasheet does not explain the units or scale.

**Comment by RobTillaart** [comment]
As the ACD3100 library is operational, I close this issue.

If there are questions or issues with respect to the ACD3100 please open a new issue here:

- https://github.com/RobTillaart/ACD3100
